### ISSUE TITLE
Mirror TLS support

### ISSUE THREAD

**Issue by steven-sheehy** [hashgraph/hedera-sdk-java#494]
Currently the `Client` only provides a list of node endpoints that include the insecured port. Providing a custom set of endpoints that includes the secure port 50212 still fails because the `ManagedNode` hardcodes `plaintext` for the gRPC channel. This also occurs for the mirror node. The mirror node will soon be transitioning to endpoints that will only offer a secured port 443 so we need TLS support very soon.

It would be nice if there was some option that a client could separately configure mirror and main nodes to turn on TLS. If enabled, it would default to port 50212 instead of 50211 for the existing `Client.for()` helper methods. It would also set plaintext on the channel only if this TLS option is disabled. For mirror nodes it could change default port from 5600 to 443. However, the client should still be allowed to set a non-standard port and still set this TLS flag.

**Comment by danielakhterov** [comment]
@steven-sheehy A few questions:
 - Is testnet updated to support TLS over 50212? 
 - Is this TLS v1.3 only; _iirc_ this is the case which will make it very difficult to support in the other SDK's.
 - Where are the certs? _IIRC_ they're in some file on the Hedera network, but I don't know which file, and what is the format of that file.

**Comment by steven-sheehy** [comment]
> * Is testnet updated to support TLS over 50212?

Yes:

```
openssl s_client -showcerts -connect 0.testnet.hedera.com:50212
CONNECTED(00000005)
depth=0 C = US, ST = TX, L = Richardson, O = Swirlds, OU = Hedera Mainnet, CN = 00000000, emailAddress = admin@swirlds.com
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = US, ST = TX, L = Richardson, O = Swirlds, OU = Hedera Mainnet, CN = 00000000, emailAddress = admin@swirlds.com
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/C=US/ST=TX/L=Richardson/O=Swirlds/OU=Hedera Mainnet/CN=00000000/emailAddress=admin@swirlds.com
   i:/C=US/ST=TX/L=Richardson/O=Swirlds/OU=Hedera Mainnet/CN=00000000/emailAddress=admin@swirlds.com
-----BEGIN CERTIFICATE-----
MIIE5TCCA02gAwIBAgIUFTt/PHfHVCRde4wwkTbqX5fvaLswDQYJKoZIhvcNAQEM
BQAwgY8xCzAJBgNVBAYTAlVTMQswCQYDVQQIDAJUWDETMBEGA1UEBwwKUmljaGFy
ZHNvbjEQMA4GA1UECgwHU3dpcmxkczEXMBUGA1UECwwOSGVkZXJhIE1haW5uZXQx
ETAPBgNVBAMMCDAwMDAwMDAwMSAwHgYJKoZIhvcNAQkBFhFhZG1pbkBzd2lybGRz
LmNvbTAgFw0yMDA0MjAyMDIxMDJaGA8yMjk0MDIwMjIwMjEwMlowgY8xCzAJBgNV
BAYTAlVTMQswCQYDVQQIDAJUWDETMBEGA1UEBwwKUmljaGFyZHNvbjEQMA4GA1UE
CgwHU3dpcmxkczEXMBUGA1UECwwOSGVkZXJhIE1haW5uZXQxETAPBgNVBAMMCDAw
MDAwMDAwMSAwHgYJKoZIhvcNAQkBFhFhZG1pbkBzd2lybGRzLmNvbTCCAaIwDQYJ
KoZIhvcNAQEBBQADggGPADCCAYoCggGBAOXnle5S3POX/vIIQLvqR20kfZ6hYI5s
5Kcibb8dEP8TgrcvPKoB4glquxVZxCw8C1pEQ5uLjRmmK8cEE/HNnXsr71CDmZoI
v3cWCrF1MkehLQIF+KMdqfgf9PIUv7pmgKcc5hTlPEResdUejEIC0cMjhFrAfZq5
Z1Wzeci5PMpo6op/QhpoOPwPl0k4Ljm8pxKeIAj6edsm+LP50vnMo8v5GWc7k7tI
QiI3bO+P/H0bN4j4t/OS0scG9bG/SnrsQv4MBnUr+zwpaaCKvRUF53klSE6N6ZVH
BBVB+cLWpXO45loXMzuCLm1fTZxxwx04eRBAUylnNEXF+As1eqvZrPt1UCD1UpKi
+VvkkZlPDu9iUAI31/NaoXym0Gz4LIbpXG7NR017V82TRxQi6Q4WGJhqWek0Ytm+
wZn00CIb/V8ZRyojxdlIAlXXWVOU3LzrbuK9QOTdwN1dt3IVEdZ6EV73z75GrL2W
xBTQZdwd/BM6jOJuvVWbEBObZ33G7kcKfwIDAQABozUwMzAPBgNVHREECDAGhwR/
AAABMAsGA1UdDwQEAwIEMDATBgNVHSUEDDAKBggrBgEFBQcDATANBgkqhkiG9w0B
AQwFAAOCAYEA3PLAymODqCJ7Q8NnTB3De9onQV7h+9Q2TByIoeOiKwhKvX3R6LTy
cwChTAOyaS7hH5i6JXOzgQac0g7teFKH4LPemTfRuUH7p0CmbaJe2oZOWjxjI3aZ
SBRhDi2qz9ehGZBHxqtNhezm5cbxHBW3fm80FSeIDM/mt028Y5Vb2j/Y+7j/BwY4
4qHL5wAS2qb8qM2Ki5c0DVdJBszl9QeiLYk1RgrIC0gMLTAs7KO7TGDUbgAsskb4
eSLvnSgA761irNgtl5p4VzZR51ft7icRYGShG+GPMV+ZvFtRGiamzYHZeDt3Ehky
oqW+217QVgelnZ8sI+BZAP0RWnQwKN4St8rPLjppvXTDE4Vz0C21WSYqtsz/hasV
iIu+WMhb0fwdjvWDe5Y48MLWWkxsHcjI7WW6ZVgtSsL7j1Wbb4QIZBMEK66K0QFc
Mpy/h47JAuZeAyMa66ArFfCAa6QwLMqLTeeNS4NWRFwDstSFmUiuizap8QQvafXa
jMbYW4ufxHA6
-----END CERTIFICATE-----
---
Server certificate
subject=/C=US/ST=TX/L=Richardson/O=Swirlds/OU=Hedera Mainnet/CN=00000000/emailAddress=admin@swirlds.com
issuer=/C=US/ST=TX/L=Richardson/O=Swirlds/OU=Hedera Mainnet/CN=00000000/emailAddress=admin@swirlds.com
---
No client certificate CA names sent
Server Temp Key: DH, 1024 bits
---
SSL handshake has read 2063 bytes and written 386 bytes
---
New, TLSv1/SSLv3, Cipher is DHE-RSA-AES256-GCM-SHA384
Server public key is 3072 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : DHE-RSA-AES256-GCM-SHA384
    Session-ID: 6313096D15D232B75A9DB955751970D73D192DF4965F7F818046D5813390DC88
    Session-ID-ctx:
    Master-Key: 19AD4A8D4F508298E7BF3F5A37EDF3235A4044EB7CF863D1D14D89280E62A521FCE8DCAA9924F5DADF05743AE0DA1C13
    Start Time: 1621520818
    Timeout   : 7200 (sec)
    Verify return code: 21 (unable to verify the first certificate)
---
closed
```

> * Is this TLS v1.3 only; _iirc_ this is the case which will make it very difficult to support in the other SDK's.

No, I see TLSv1.2 above.

> * Where are the certs? _IIRC_ they're in some file on the Hedera network, but I don't know which file, and what is the format of that file.

See above. Server certificates are retrieved automatically as part of the TLS protocol. You don't need to configure the client with any certificates to make it work unless the server is using self-signed certs (which we're not) or there's mutual authentication (which we're not). The Java/gRPC toolchain can already pull the server certs and verify them against its internal well-known certificate authorities.

To make the SDK work against the TLS endpoint, all that needs to be done is to use the correct port and change the channel to not set plaintext. I've already tested locally by hacking `ManagedNode` and it works against mainnet TLS endpoints.

**Comment by mehcode** [comment]
Previously Hedera had very strict cipher requirements which is why this was not done. As the required API to configure those is not available in the gRPC lib that we have to use and would require a ton of class reflection. If those requirements are gone, this is stupidly trivial and we can get it done pretty quickly.

**Comment by danielakhterov** [comment]
@steven-sheehy
Firstly, thank you for the in-depth response, I really appreciate it. However, I just tested some integration tests with TLS, and I'm getting `SSLV3_ALERT_HANDSHAKE_FAILURE` for both `mainnet` and `testnet`. I've changed the port to `50212` and updated `ManagedNode` to not use the `usePlainText()` method. I've also tried with `.useTransportLayerSecurity()`, but that seems to be on by default since I get the same error back. Perhaps I'm forgetting to do something.

**Comment by steven-sheehy** [comment]
What version is your JRE? Different JREs have different TLS support. The main nodes only [support](https://github.com/hashgraph/hedera-services/blob/020bdfe1300a83a42391c265e6fc153eeec163b5/hedera-node/src/main/java/com/hedera/services/grpc/ConfigDrivenNettyFactory.java#L50) TLSv1.2+. You're attempting to send a SSLv3 Hello and the server is denying it. SSLv3 and TLSv1.1 have been removed from most JRE's since they're insecure.

Since the SDK supports Java 7+, it's possible that we can't support TLS for those older JREs. We can just document what the minimum JRE required is to use the TLS option.

**Comment by danielakhterov** [comment]
```
❯ java --version
openjdk 14.0.2 2020-07-14
OpenJDK Runtime Environment (build 14.0.2+12-46)
OpenJDK 64-Bit Server VM (build 14.0.2+12-46, mixed mode, sharing)
```

@mehcode and I talked about his more, and the only thing we can think of is maybe the SDK needs to load the proper channel builder from `grpc-netty-shaded` or `okhttp` depending on which one is being used.

**Comment by steven-sheehy** [comment]
Narrowing scope to just add TLS support for mirror node. Mirror node uses a CA signed cert. Any client API that results should take into consideration future support for main node TLS and design for consistency.

**Comment by mehcode** [comment]
This was released with the recent 2.0.6 (basic support for explicit :443 or :50212). We'll figure out an actual API for TLS once the issues with HAPI nodes is resolved.
